src/inmanta/loader.py

```diff
@@ -184,69 +133,57 @@ def get_object_source(self, instance: object) -> Optional[str]:
 
     def get_file_hashes(self) -> Iterable[str]:
         """Return the hashes of all source files"""
-        return (info.hash for info in self.__file_info.values())
-
-    def get_file_content(self, hash: str) -> bytes:
-        """Get the file content for the given hash"""
-        for info in self.__file_info.values():
-            if info.hash == hash:
-                return info.content
-
-        raise KeyError("No file found with this hash")
-
-    def get_types(self) -> Iterable[tuple[str, list[SourceInfo]]]:
-        """Get a list of all registered types"""
-        return ((type_name, [self.__file_info[path] for path in files]) for type_name, files in self.__type_file.items())
-
+        return (info.meta_data.hash_value for info in self.__file_info.values())
+
+    def get_inmanta_module_requirements(self, module_name: str) -> set[str]:
+        """Get the list of python requirements associated with this inmanta module"""
+        project: module.Project = module.Project.get()
+        mod: module.Module = project.modules[module_name]
+        if project.metadata.agent_install_dependency_modules:
+            _requires = mod.get_all_python_requirements_as_list()
+        else:
+            _requires = mod.get_strict_python_requirements_as_list()
+        return set(_requires)
 
-@dataclass(frozen=True)
-@functools.total_ordering
-class ModuleSource:
-    """
-    :param name: the name of the python module. e.g. inmanta_plugins.model.x
-    :param is_byte_code: is this content python byte code or python source
-    :param source: the content of the file
-    :param _client: a protocol client, required when source is not set
+    def get_module_version_info(self) -> dict[str, "InmantaModuleDTO"]:
+        """Return all module version info"""
 
-    """
+        modules_data = {}
+        for module_name, module_sources in self.__module_to_source_info.items():
 
-    name: str
-    hash_value: str
-    is_byte_code: bool
-    source: Optional[bytes] = None
-    _client: Optional["protocol.SyncClient"] = None
+            requirements = self.get_inmanta_module_requirements(module_name)
+            files_metadata: list[ModuleSourceMetadata] = [module_source.meta_data for module_source in module_sources]
 
-    def __lt__(self, other):
-        if not isinstance(other, ModuleSource):
-            return NotImplemented
-        return (self.name, self.hash_value, self.is_byte_code) < (other.name, other.hash_value, other.is_byte_code)
+            module_version = self.get_module_version(requirements, files_metadata)
 
-    def __eq__(self, other: object) -> bool:
-        if not isinstance(other, ModuleSource):
-            return False
-        return (self.name, self.hash_value, self.is_byte_code) == (other.name, other.hash_value, other.is_byte_code)
+            modules_data[module_name] = InmantaModuleDTO(
```

Comment:
The logic that leads here looks good, but I wonder if perhaps what was there before lingered a bit too much? You now have all these intermediate fields (the `__` ones), that used to be built incrementally (iirc), but are now really built on a per-module basis already, so why not simply keep this here in an attribute instead? i.e. construct it in `register_inmanta_module`, store it there, get rid of `__module_to_source_info` and `__module_to_agents` and a bunch of methods. And perhaps even `__file_info`. I think it would all become quite a bit more cohesive.

Reply:
iirc the rationale at the time to keep doing it this way was that InmantaModuleDTO was frozen (i.e. had the attribute `model_config: ClassVar[ConfigDict] = ConfigDict(frozen=True)` so I couldn't create the object once and update it.
Unless I'm missing something, I can't find a good reason for it to be frozen anymore so I will update this part.